src/OpenTelemetry/Trace/ActivityProcessor.cs

```diff
@@ -13,16 +13,19 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 // </copyright>
+
+using System;
 using System.Diagnostics;
 using System.Threading;
 using System.Threading.Tasks;
+using OpenTelemetry.Internal;
 
 namespace OpenTelemetry.Trace
 {
     /// <summary>
     /// Activity processor base class.
     /// </summary>
-    public abstract class ActivityProcessor
+    public abstract class ActivityProcessor : IDisposable
```

Comment:
do we want to conditionally implement IAsyncDisposable for netstandard2.1?

Reply:
we could. should we add netstandard2.1 as support to try that?

Reply:
I think we should. Probably create an issue for tracking and resolve in anther PR.

Reply:
I was looking at the docs, but it's not very clear for me how that works.
When we implement IDiposable (https://docs.microsoft.com/en-us/dotnet/api/system.idisposable?view=netcore-3.1#examples), we need:
public void Dispose() and protected virtual void Dispose(bool disposing)

When we implement IAsyncDisposable (https://docs.microsoft.com/en-us/dotnet/api/system.iasyncdisposable?view=dotnet-plat-ext-3.1), we need:
public ValueTask DisposeAsync()

Looking for some examples, I saw people calling the Dispose(), something like this:
```csharp
public virtual ValueTask DisposeAsync()
        {
                Dispose();
                return default;
        }
```

What do you think?

Reply:
I don't think you want to call Dispose from DisposeAsync. You don't want to be doing stuff like GetAwaiter().GetResult in this fully async path. I would probably keep the dispose paths separate. You would only call one or the other. I have this in a base class in my own codebase:

```
        /// <inheritdoc/>
        public void Dispose()
        {
            if (Interlocked.CompareExchange(ref this.disposed, 1, 0) == 1)
            {
                return;
            }

            this.Dispose(true);
            GC.SuppressFinalize(this);
        }

#if NETSTANDARD2_1
        /// <inheritdoc/>
        public async ValueTask DisposeAsync()
        {
            if (Interlocked.CompareExchange(ref this.disposed, 1, 0) == 1)
            {
                return;
            }

            // see https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-disposeasync#the-disposeasync-method
            // Perform async cleanup.
            await this.DisposeAsyncCore().ConfigureAwait(false);

            // Dispose of managed resources.
            this.Dispose(false);

            // Suppress finalization.
            GC.SuppressFinalize(this);
        }
#endif

        /// <summary>
        /// Releases unmanaged and - optionally - managed resources.
        /// </summary>
        /// <param name="disposing"><c>true</c> to release both managed and unmanaged resources; <c>false</c> to release only unmanaged resources.</param>
        protected virtual void Dispose(bool disposing)
        {
        }

#if NETSTANDARD2_1
        /// <summary>
        /// Releases unmanaged and - optionally - managed resources asynchronously.
        /// </summary>
        /// <returns>a ValueTask</returns>
        protected virtual ValueTask DisposeAsyncCore()
        {
            return default;
        }
#endif
    }
```

Reply:
should i add this in or should i create another one after this is merged?

Reply:
Not in this PR.

Reply:
i will open a new issue and we can discuss this there! 👍

Reply:
https://github.com/open-telemetry/opentelemetry-dotnet/issues/982